### PR TITLE
simple code of data scrape

### PR DESCRIPTION
**Pull request by ZiaurRehmanmalhi**

### CODE REVIEW

File: tasks/task_2.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+driver.get(url)
+last_height = driver.execute_script("return document.documentElement.scrollHeight")
+
+while True:
+    driver.execute_script("window.scrollTo(0, arguments[0]);", last_height)
+    time.sleep(1.5)
+    new_height = driver.execute_script("return document.documentElement.scrollHeight")
+    if new_height == last_height:
+        break
+    last_height = new_height
```
Comment (Hamma111) [inline tasks/task_2.py]: Can we create a generic function out of this `def scroll_to_page_end(driver)` which will take driver as parameter and scroll the opened page to the end of page?

Move this function to a `utils.py` file, and import, and call that here.

File: tasks/task_2.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+html = driver.page_source
+soup = BeautifulSoup(html, 'html.parser')
+all_content = soup.find_all('ytd-rich-item-renderer')
```
Comment (Hamma111) [inline tasks/task_2.py]: `video_elements` would be a better name

File: tasks/task_2.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+data = []
+for videos in all_content:
+    thumbnail = videos.find('img', {"style": "background-color: transparent;"}).get('src')
+    title = videos.find('yt-formatted-string', {"id": "video-title"}).text
+    views = videos.find('span', {"class": "inline-metadata-item style-scope ytd-video-meta-block"}).text
+    upload_time = videos.find('span', {"class": "inline-metadata-item style-scope ytd-video-meta-block"}).get('before')
+    video_duration = videos.find(
+        'span', {"class": "style-scope ytd-thumbnail-overlay-time-status-renderer"}
+    ).text.strip()
```
Comment (Hamma111) [inline tasks/task_2.py]: Scrape and store `uploaded_at` as well

File: tasks/task_2.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    all_data = {
+        "Title": title,
+        "Views": views,
+        "Video Duration": video_duration,
+        "Thumbnail Link": thumbnail
+    }
+    data.append(all_data)
+
+df = pd.DataFrame(data)
+df.to_csv('youtube_data.csv', index=False)
```
Comment (Hamma111) [inline tasks/task_2.py]: the filename should be the channel's name which is. you can extract that `driver.current_url` or `driver.title`. 

In this case filename should be `zusmani78.csv`

File: tasks/task_2.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    upload_time = videos.find('span', {"class": "inline-metadata-item style-scope ytd-video-meta-block"}).get('before')
+    video_duration = videos.find(
+        'span', {"class": "style-scope ytd-thumbnail-overlay-time-status-renderer"}
+    ).text.strip()
+ 
+    all_data = {
+        "Title": title,
+        "Views": views,
+        "Video Duration": video_duration,
+        "Thumbnail Link": thumbnail
```
Comment (Hamma111) [inline tasks/task_2.py]: why are you not storing `upload_time`?

File: tasks/task_2.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    data.append(all_data)
```
Comment (Hamma111) [inline tasks/task_2.py]: Move all the scraping logic into a function `def extract_videos_data(soup)`. This function should return the list of dictionaries which is `data` in your case. 

All scraping logic should be in a file `scraper.py`. import and use wherever needed.

File: tasks/task_2.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+df = pd.DataFrame(data)
+df.to_csv('youtube_data.csv', index=False)
+print("Data saved to csv file")
```
Comment (Hamma111) [inline tasks/task_2.py]: Your main file which is task_2.py should be as thin and small as possible.

**Review by Hamma111 — CHANGES_REQUESTED** [review]

File: tasks/task_2.py (hunk 1 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 2 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 6 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 7 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 3 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 5 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: sir views or upload_time ka html class same thi iss wja se bgt koshish ki lekin nhi aya main div ko pkr k split b kia lekin ab b koshish kr rha hn jld solve ho jae ga

File: tasks/task_2.py (hunk 6 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 4 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 7 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: tasks/task_2.py (hunk 3 above)
Comment (ZiaurRehmanmalhi) [reply tasks/task_2.py]: ok

File: .idea/inspectionProfiles/profiles_settings.xml
Comment (Hamma111) [inline .idea/inspectionProfiles/profiles_settings.xml:1]: how many times to we do go over this that useless files must not be committed?

Comment (Hamma111) [reply .idea/inspectionProfiles/profiles_settings.xml:1]: Always commit the relevant files only.

File: tasks/Task_2.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+from selenium import webdriver
+from bs4 import BeautifulSoup
+import pandas as pd
+from utils import scroll_to_page_end, extract_videos_data
+
+url = "https://www.youtube.com/@zusmani78/videos"
+driver = webdriver.Chrome(executable_path="chromedriver")
+driver.get(url)
+scroll_to_page_end(driver)       # utils function_1 call
```
Comment (Hamma111) [inline tasks/Task_2.py]: no need for comments, code is self-explanatory

File: tasks/Task_2.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+html = driver.page_source
+soup = BeautifulSoup(html, 'html.parser')
+data = extract_videos_data(soup)      # utils function_2 call
+
+dataframe = pd.DataFrame(data)
+dataframe.to_csv('zusmani78.csv', index=False)
```
Comment (Hamma111) [inline tasks/Task_2.py]: Dont hard-code the channel name, extract it from the URL or from page source somewhere

**Review by Hamma111 — APPROVED** [review]

**Pull request merged**